Mathlib/Algebra/Homology/ShortComplex/LeftHomology.lean

```diff
@@ -61,6 +60,206 @@ structure LeftHomologyData where
+
+variable (S)
+
+/-- When the second map `S.g` is zero, this is the left homology data on `S` given
+by any colimit cokernel cofork of `S.f` -/
+@[simps]
+def ofIsColimitCokernelCofork (hg : S.g = 0) (c : CokernelCofork S.f) (hc : IsColimit c) :
+    S.LeftHomologyData where
+  K := S.X₂
+  H := c.pt
+  i := 𝟙 _
+  π := c.π
+  wi := by rw [id_comp, hg]
+  hi := KernelFork.IsLimit.ofId _ hg
+  wπ := CokernelCofork.condition _
+  hπ := IsColimit.ofIsoColimit hc (Cofork.ext (Iso.refl _) (by aesop_cat))
+
+@[simp] lemma ofIsColimitCokernelCofork_f' (hg : S.g = 0) (c : CokernelCofork S.f)
+    (hc : IsColimit c) : (ofIsColimitCokernelCofork S hg c hc).f' = S.f := by
+  rw [← cancel_mono (ofIsColimitCokernelCofork S hg c hc).i, f'_i,
+    ofIsColimitCokernelCofork_i]
+  dsimp
+  rw [comp_id]
+
+/-- When the second map `S.g` is zero, this is the left homology data on `S` given by
+the chosen `cokernel S.f` -/
+@[simps!]
+noncomputable def ofHasCokernel [HasCokernel S.f] (hg : S.g = 0) : S.LeftHomologyData :=
+  ofIsColimitCokernelCofork S hg _ (cokernelIsCokernel _)
+
+/-- When the first map `S.f` is zero, this is the left homology data on `S` given
+by any limit kernel fork of `S.g` -/
+@[simps]
+def ofIsLimitKernelFork (hf : S.f = 0) (c : KernelFork S.g) (hc : IsLimit c) :
+    S.LeftHomologyData where
+  K := c.pt
+  H := c.pt
+  i := c.ι
+  π := 𝟙 _
+  wi := KernelFork.condition _
+  hi := IsLimit.ofIsoLimit hc (Fork.ext (Iso.refl _) (by aesop_cat))
+  wπ := Fork.IsLimit.hom_ext hc (by
+    dsimp
+    simp only [comp_id, zero_comp, Fork.IsLimit.lift_ι, Fork.ι_ofι, hf])
+  hπ := CokernelCofork.IsColimit.ofId _ (Fork.IsLimit.hom_ext hc (by
+    dsimp
+    simp only [comp_id, zero_comp, Fork.IsLimit.lift_ι, Fork.ι_ofι, hf]))
+
+@[simp] lemma ofIsLimitKernelFork_f' (hf : S.f = 0) (c : KernelFork S.g) (hc : IsLimit c) :
+    (ofIsLimitKernelFork S hf c hc).f' = 0 := by
+  rw [← cancel_mono (ofIsLimitKernelFork S hf c hc).i, f'_i, hf, zero_comp]
+
+/-- When the first map `S.f` is zero, this is the left homology data on `S` given
+by the chosen `kernel S.g` -/
+@[simp]
+noncomputable def ofHasKernel [HasKernel S.g] (hf : S.f = 0) : S.LeftHomologyData :=
+  ofIsLimitKernelFork S hf _ (kernelIsKernel _)
+
+/-- When both `S.f` and `S.g` are zero, the middle object `S.X₂` gives a left homology data on S -/
+@[simps]
+def ofZeros (hf : S.f = 0) (hg : S.g = 0) : S.LeftHomologyData where
+  K := S.X₂
+  H := S.X₂
+  i := 𝟙 _
+  π := 𝟙 _
+  wi := by rw [id_comp, hg]
+  hi := KernelFork.IsLimit.ofId _ hg
+  wπ := by
+    change S.f ≫ 𝟙 _ = 0
+    simp only [hf, zero_comp]
+  hπ := CokernelCofork.IsColimit.ofId _ hf
+
+@[simp] lemma ofZeros_f' (hf : S.f = 0) (hg : S.g = 0) :
+    (ofZeros S hf hg).f' = 0 := by
+  rw [← cancel_mono ((ofZeros S hf hg).i), zero_comp, f'_i, hf]
+
+end LeftHomologyData
+
+/-- A short complex `S` has left homology when there exists a `S.LeftHomologyData` -/
+class HasLeftHomology : Prop where
```

Comment:
This class is mentioned as `TODO` in the module docstring at the top of the file. Should that docstring now be updated?